Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+{-# OPTIONS_GHC -Wall #-}
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/04-typeclasses.pdf
+
+module Week4.HW04 where
+
+import Data.List (dropWhileEnd)
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+instance (Num a, Eq a)              => Eq (Poly a) where
+    -- This will not be effecient way of doing it I think,
+    -- because its taking extra storage to save the new list 
+    -- without trailing zeros and then does the matching
+    P a == P b                      = withoutTrailingZerosA == withoutTrailingZerosB
+        where 
+            withoutTrailingZerosA   = dropWhileEnd (== 0) a 
+            withoutTrailingZerosB   =  dropWhileEnd (== 0) b
+        
+
+-- Exercise 3 -----------------------------------------
+
+-- e == exponent 
+-- c == count of the polynomial term, is it first, second, third and so on
+showHelper :: (Num a, Show a, Eq a) => Poly a -> Int -> Int -> String
+showHelper (P (p:ps)) e c
+    | p == 0                = showHelper (P ps) (e-1) (c+1)
```

Comment:
The way i have implemented the code for show, doing this(removing `dropWhileEnd`) would give me errors on edge cases like 
**notice the plus in the beginning** 

<img width="325" alt="Screenshot 2021-01-26 at 2 59 41 PM" src="https://user-images.githubusercontent.com/30658517/105826782-22270a00-5fe7-11eb-8126-d4f86682a217.png">

`dropWhileEnd` prevents things like this from happening. (i wouldnt know if the term i am printing is the first term or not after recusively going through a 100 zeros, i  wont know if all the terms i looped through are zeros and this non-zero one doesnt need to have a '+' in front of it because its the first term i am about to print) also this will prevent unnecessary recursion for those 100 zeros 

The `p==0` is necessary to handle the internal zeros if they are not leading or trailing but somewhere in the middle of the term

The logic I agree is scattered in two functions and i could have done better than to do it this way but lemme know if these are non-valid reasons or do they make sense ?

Reply:
Ok, it makes sense the way it is. But it seems to me that using zipWith to get access to current index could get rid of some of the statefulness of this implementation and simplify things somewhat.